Updates to WebServer, ChakraHost and StartupTask

gloveboxes opened the pull request:
Updates to Web Server - improved network reliability 
StartupTask - commented out StartMostRecentScript and Save calls
ChakraHost - changed scripting processing strategy to avoid race conditons

gloveboxes committed:
Updates to WebServer, ChakraHost and StartupTask

msftclas commented:
Hi **@gloveboxes**, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
**@gloveboxes**, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

gloveboxes committed:
User Guide

alecont commented on IoTBlockly/ChakraHost/ChakraHost.csproj:
```diff
@@ -11,7 +11,7 @@
     <AssemblyName>ChakraHost</AssemblyName>
     <DefaultLanguage>en-US</DefaultLanguage>
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
-    <TargetPlatformVersion>10.0.10586.0</TargetPlatformVersion>
```
Dave, why change the target platform?

gloveboxes replied:
no more reason than the latest SDK was released, I wanted to test it worked ok...

alecont commented on IoTBlockly/ChakraHost/ChakraHostAsync.cs:
```diff
@@ -22,20 +23,24 @@ public class ChakraHostAsync
-        private Queue taskQueue = new Queue();
-        private IAsyncAction workItem;
-        public bool running { get; private set; }
+       // private Queue taskQueue = new Queue();
```
Uhm... I guess it's ok to remove this as we don't use the continuations... but why remove this? did they create problems?

gloveboxes replied:
as below, never used.  Commented out as I felt guilty removing your code:)

alecont commented on IoTBlockly/ChakraHost/ChakraHostAsync.cs:
```diff
@@ -22,20 +23,24 @@ public class ChakraHostAsync
         private readonly ManualResetEventSlim waitEvent = new ManualResetEventSlim(false);
+        private AutoResetEvent newScriptEvent = new AutoResetEvent(false);
+
+        bool initilised = false;
```
typo

alecont commented on IoTBlockly/ChakraHost/ChakraHostAsync.cs:
```diff
@@ -51,7 +56,7 @@ public string init()
             JavaScriptPromiseContinuationCallback promiseContinuationCallback = delegate (JavaScriptValue task, IntPtr callbackState)
             {
-                taskQueue.Enqueue(task);
+               // taskQueue.Enqueue(task);
```
if we decide to remove the taskQueue, let's just remove these and not just comment out. I wonder if this queue gets populated ever. Can you check?

gloveboxes replied:
I did check (a lot) and it's never called

gloveboxes replied:
will remove - as above felt a little guilty removing your code:)

alecont commented on IoTBlockly/ChakraHost/ChakraHostAsync.cs:
```diff
@@ -90,109 +96,69 @@ public string dispose()
 
-        public void runScript(string script)
+
+        public void ScriptProcessor()
```
I like the new ScriptProcessor(). Few comments:
- minor: let's use the same "scriptProcessor()" naming convention to be consistent
- where do we call init()? I don't see a call to init() anymore...

gloveboxes replied:
Ah capitalised following C# method naming conventions, but am easy:)  init() is called inside the try catch block after the waitone.

alecont commented on IoTBlockly/IoTBlocklyBackgroundApp/IoTBlocklyBackgroundApp.csproj:
```diff
@@ -11,7 +11,7 @@
     <DefaultLanguage>en-US</DefaultLanguage>
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
-    <TargetPlatformVersion>10.0.10586.0</TargetPlatformVersion>
+    <TargetPlatformVersion>10.0.14393.0</TargetPlatformVersion>
```
same as above

gloveboxes replied:
as above - more for testing as the 19393 SDK was final release.

alecont commented on IoTBlockly/IoTBlocklyBackgroundApp/StartupTask.cs:
```diff
@@ -34,7 +34,7 @@ public sealed class StartupTask : IBackgroundTask
             var publicFolder = await localFolder.GetFolderAsync("public");
 
-            await StartMostRecentScript();
+            // await StartMostRecentScript();  // problematic if JS Corrupt when stored
```
Let's add a REVIEW comment to improve this and that is ok to comment this out

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -192,6 +193,7 @@ private Task WriteHeaderNoFlushAsync(Stream resp, int statusCode, string message
                 string header = String.Format(
                     "HTTP/1.1 {0} {1}\r\n" +
                     location +
+                    "Cache-Control:public, max-age=31536000\r\n" + // randon big number to ensure browser gets from cache dglover@microsoft.com 
```
good change. no need to add your alias to the code. Git history will keep that info

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -214,8 +216,8 @@ private Task WriteHeaderNoFlushAsync(Stream resp, int statusCode, string message
             listener = new StreamSocketListener();
-            listener.Control.KeepAlive = true;
-            listener.Control.NoDelay = true;
+            //listener.Control.KeepAlive = true;
```
let's just remove these; no need to keep the comments

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -246,90 +248,104 @@ public void Post(string path, RouteCallback callback)
-                    await input.ReadAsync(buffer, BufferSize, InputStreamOptions.Partial);
-                    request.Append(Encoding.UTF8.GetString(data, 0, data.Length));
-                    dataRead = buffer.Length;
+                    while (dataRead == BufferSize)
```
I like this :-)

gloveboxes replied:
yes similar thought had crossed my mind - changing to do {} while, testing now

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -246,90 +248,104 @@ public void Post(string path, RouteCallback callback)
             {
-                while (dataRead == BufferSize)
+                // grab the request
+                StringBuilder request = new StringBuilder();
```
I like this :-)

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -246,90 +248,104 @@ public void Post(string path, RouteCallback callback)
-                    await input.ReadAsync(buffer, BufferSize, InputStreamOptions.Partial);
-                    request.Append(Encoding.UTF8.GetString(data, 0, data.Length));
-                    dataRead = buffer.Length;
+                    while (dataRead == BufferSize)
```
is there a case when dataRead != BufferSize but dataRead > 0? are we leaving some bytes on the ground?

gloveboxes replied:
thought ha crossed my mind too, I look at again

alecont replied:
any news on this? :-)

alecont replied:
checked with Dave offline, we're good here

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.cs:
```diff
@@ -246,90 +248,104 @@ public void Post(string path, RouteCallback callback)
+                        if (!handled)
                         {
-                            if (path.StartsWith(t.Item1))
+                            foreach (var t in getStaticRoutes)
```
nothing changed here right? just spaces/tabs?

alecont commented on IoTBlockly/SimpleWebServer/SimpleWebServer.csproj:
```diff
@@ -11,7 +11,7 @@
     <DefaultLanguage>en-US</DefaultLanguage>
     <TargetPlatformIdentifier>UAP</TargetPlatformIdentifier>
-    <TargetPlatformVersion>10.0.10586.0</TargetPlatformVersion>
+    <TargetPlatformVersion>10.0.14393.0</TargetPlatformVersion>
```
same comment as above

alecont commented:
Hi Dave, sorry for the delay. I added some comments. can you review?

gloveboxes committed:
Capitialised method names ChakraHost

removed commented out unused variables
removed email address from comments
Changes pause from eventwait to Task.Delay().Wait()

alecont commented:
looks good. merging.
